pkg/apis/core/validation/validation.go

```diff
@@ -4671,6 +4673,49 @@ func ValidatePodUpdate(newPod, oldPod *core.Pod, opts PodValidationOptions) fiel
 		mungedPodSpec.TerminationGracePeriodSeconds = oldPod.Spec.TerminationGracePeriodSeconds // +k8s:verify-mutation:reason=clone
 	}
 
+	// Handle validations specific to gated pods.
+	podIsGated := len(oldPod.Spec.SchedulingGates) > 0
+	if opts.AllowMutableNodeSelectorAndNodeAffinity && podIsGated {
+		// Additions to spec.nodeSelector are allowed (no deletions or mutations) for gated pods.
+		if !apiequality.Semantic.DeepEqual(mungedPodSpec.NodeSelector, oldPod.Spec.NodeSelector) {
+			allErrs = append(allErrs, validateNodeSelectorMutation(specPath.Child("nodeSelector"), mungedPodSpec.NodeSelector, oldPod.Spec.NodeSelector)...)
+			mungedPodSpec.NodeSelector = oldPod.Spec.NodeSelector // +k8s:verify-mutation:reason=clone
+		}
+
+		// Validate node affinity mutations.
+		var oldNodeAffinity *core.NodeAffinity
+		if oldPod.Spec.Affinity != nil {
+			oldNodeAffinity = oldPod.Spec.Affinity.NodeAffinity // +k8s:verify-mutation:reason=clone
+		}
+
+		var mungedNodeAffinity *core.NodeAffinity
+		if mungedPodSpec.Affinity != nil {
+			mungedNodeAffinity = mungedPodSpec.Affinity.NodeAffinity // +k8s:verify-mutation:reason=clone
+		}
+
+		if !apiequality.Semantic.DeepEqual(oldNodeAffinity, mungedNodeAffinity) {
+			allErrs = append(allErrs, validateNodeAffinityMutation(specPath.Child("affinity").Child("nodeAffinity"), mungedNodeAffinity, oldNodeAffinity)...)
+			if mungedPodSpec.Affinity != nil {
+				mungedPodSpec.Affinity.NodeAffinity = oldNodeAffinity // +k8s:verify-mutation:reason=clone
+			}
+		}
+
+		// All updates to PreferredDuringSchedulingIgnoredDuringExecution are allowed for gated pods.
+		var oldPreferredSchedulingTerms []core.PreferredSchedulingTerm
+		if oldNodeAffinity != nil {
+			oldPreferredSchedulingTerms = oldNodeAffinity.PreferredDuringSchedulingIgnoredDuringExecution
+		}
+		if len(oldPreferredSchedulingTerms) != 0 {
+			if mungedPodSpec.Affinity == nil {
+				mungedPodSpec.Affinity = &core.Affinity{
+					NodeAffinity: &core.NodeAffinity{},
+				}
+			}
+			mungedPodSpec.Affinity.NodeAffinity.PreferredDuringSchedulingIgnoredDuringExecution = oldPreferredSchedulingTerms // +k8s:verify-mutation:reason=clone
+		}
```

Comment:
none of this should be needed, since we set `mungedPodSpec.Affinity.NodeAffinity = oldNodeAffinity` above, right?

Reply:
Ah, I was using this to handle the test case `preferredDuringSchedulingIgnoredDuringExecution can have removals for gated pods`, however, now that you mention it, it _should_ have been handled by the line `mungedPodSpec.Affinity.NodeAffinity = oldNodeAffinity` above, however, that line was only executed when `mungedPodSpec.Affinity != nil`, and in this test case the new pod update had the affinity field as nil, so the line wasn't executed. 

I updated this to also still do `mungedPodSpec.Affinity.NodeAffinity = oldNodeAffinity` even if new affinity is nil, removed the highlighted code above, and all tests are passing.